css/css-view-transitions/snapshot-containing-block-absolute.html is an ImageOnlyFailure

mattwoodrow opened the pull request:
#### 01f6d0726a7f6a35c275d109126e3cac23f9c61e
<pre>
css/css-view-transitions/snapshot-containing-block-absolute.html is an ImageOnlyFailure
<a href="https://bugs.webkit.org/show_bug.cgi?id=292970">https://bugs.webkit.org/show_bug.cgi?id=292970</a>
&lt;<a href="https://rdar.apple.com/151279896">rdar://151279896</a>&gt;

Reviewed by Tim Nguyen.

These tests change the position property on the ::view-transition pseudo, and
break code that expects the normal layer structure.

Create a new anonymous renderer to wrap the view transition pseudo tree, to
represent the view transition snapshot containing block. This is always
position:fixed, and can&apos;t be styled by the author.

It&apos;s currently sized to match the RenderView, but can be easily extended to
account for insets/outsets to fix bug 285400.

Most code that previously used the view transition root changed over to use the
view transition containing block, as they generally just want the root of the VT
generated tree.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::isAnonymousBlock const):
(WebCore::RenderElement::mayContainOutOfFlowPositionedObjects const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldBeCSSStackingContext const):
(WebCore::RenderLayer::rebuildZOrderLists):
(WebCore::RenderLayer::collectLayers):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::childForSuperlayers const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateBackingAndHierarchy):
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
(WebCore::RenderLayerCompositor::fixedLayerIntersectsViewport const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isViewTransitionContainingBlock const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionContainingBlock const):
(WebCore::RenderView::setViewTransitionContainingBlock):
(WebCore::RenderView::viewTransitionRoot const): Deleted.
(WebCore::RenderView::setViewTransitionRoot): Deleted.
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::renderer const):

Canonical link: <a href="https://commits.webkit.org/295083@main">https://commits.webkit.org/295083@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48de05814e8b9a589657aa5688a7d3b3768e8ea9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103714 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13736 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108904 "Failed to checkout and rebase branch from PR 45350") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54366 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31962 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/108904 "Failed to checkout and rebase branch from PR 45350") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93560 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/108904 "Failed to checkout and rebase branch from PR 45350") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11608 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53742 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88072 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; 1 api test failed or timed out; re-run-api-tests (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111294 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30870 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22763 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87801 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31231 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87451 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22324 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32332 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10059 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25227 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30798 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36101 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30592 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33927 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32153 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -619,7 +619,7 @@ bool RenderLayer::shouldBeNormalFlowOnly() const
 bool RenderLayer::shouldBeCSSStackingContext() const
 {
-    return !renderer().style().hasAutoUsedZIndex() || renderer().shouldApplyLayoutContainment() || renderer().shouldApplyPaintContainment() || renderer().requiresRenderingConsolidationForViewTransition() || renderer().isRenderViewTransitionCapture() || renderer().isViewTransitionRoot() || isRenderViewLayer();
+    return !renderer().style().hasAutoUsedZIndex() || renderer().shouldApplyLayoutContainment() || renderer().shouldApplyPaintContainment() || renderer().requiresRenderingConsolidationForViewTransition() || renderer().isRenderViewTransitionCapture() || renderer().isViewTransitionContainingBlock() || isRenderViewLayer();
```
Is this change correct? the spec says the ::view-transition pseudo should generate a stacking context

mattwoodrow replied:
I suspect the spec just wants the VT tree as a whole to be stacking (so that it can't interleave with anything else outside of it).

The spec does explicitly say the `::view-transition` pseudo though, and it could be observable (use author styles to remove other causes of being a stacking context, make a descendant a sibling in paint order, play with z-index).

I'll make them both stacking, which I think complies with the spec and works correctly.

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h:
```diff
@@ -41,7 +41,7 @@ class RenderTreeUpdater::ViewTransition {
     void updatePseudoElementTree(RenderElement&, StyleDifference minimalStyleDifference);
 private:
-    void buildPseudoElementGroup(const AtomString&, RenderElement&, RenderObject* = nullptr);
+    void buildPseudoElementGroup(RenderBox& viewTransitionRoot, const AtomString&, RenderElement&, RenderObject* = nullptr);
```
shouldn't this be a block and not a box?

mattwoodrow replied:
I guess it can (even RenderBlockFlow, need to downcast either way), since we unconditionally use that for `::view-transition`. I think that's somewhat incorrect, since the author could have provided a `display: inline` style, but that's tangential.

alanbaradlay commented on Source/WebCore/rendering/RenderView.cpp:
```diff
@@ -1132,14 +1132,14 @@ SingleThreadWeakHashSet<RenderCounter> RenderView::takeCountersNeedingUpdate()
 }
 
-SingleThreadWeakPtr<RenderBlockFlow> RenderView::viewTransitionRoot() const
+SingleThreadWeakPtr<RenderBlockFlow> RenderView::viewTransitionContainingBlock() const
```
Any particular reason for this to be a weak and not a checked ptr?

mattwoodrow replied:
No, just existing code. I can fix it now though.

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -75,16 +76,34 @@ void RenderTreeUpdater::ViewTransition::updatePseudoElementTree(RenderElement& d
 
+    WeakPtr viewTransitionContainingBlock = documentElementRenderer.view().viewTransitionContainingBlock();
+    if (!viewTransitionContainingBlock) {
+        auto cbStyle = RenderStyle::createAnonymousStyleWithDisplay(documentElementRenderer.view().style(), DisplayType::Block);
```
we've been trying to not use abbreviations like this.

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -75,16 +76,34 @@ void RenderTreeUpdater::ViewTransition::updatePseudoElementTree(RenderElement& d
+        // FIXME: Bug 285400 - Correctly account for scrollbars and insets.
+        cbStyle.setTop(Length(0, LengthType::Fixed));
+        cbStyle.setLeft(Length(0, LengthType::Fixed));
+        cbStyle.setWidth(Length(100, LengthType::Percent));
```
this is probably redundant as blocks normally take up all the space available in inline direction, but I'd use inset { 0 } instead of percent values here.

mattwoodrow committed:
css/css-view-transitions/snapshot-containing-block-absolute.html is an ImageOnlyFailure
https://bugs.webkit.org/show_bug.cgi?id=292970
<rdar://151279896>

Reviewed by Tim Nguyen.

These tests change the position property on the ::view-transition pseudo, and
break code that expects the normal layer structure.

Create a new anonymous renderer to wrap the view transition pseudo tree, to
represent the view transition snapshot containing block. This is always
position:fixed, and can't be styled by the author.

It's currently sized to match the RenderView, but can be easily extended to
account for insets/outsets to fix bug 285400.

Most code that previously used the view transition root changed over to use the
view transition containing block, as they generally just want the root of the VT
generated tree.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::isAnonymousBlock const):
(WebCore::RenderElement::mayContainOutOfFlowPositionedObjects const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldBeCSSStackingContext const):
(WebCore::RenderLayer::rebuildZOrderLists):
(WebCore::RenderLayer::collectLayers):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::childForSuperlayers const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateBackingAndHierarchy):
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
(WebCore::RenderLayerCompositor::fixedLayerIntersectsViewport const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isViewTransitionContainingBlock const):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::viewTransitionContainingBlock const):
(WebCore::RenderView::setViewTransitionContainingBlock):
(WebCore::RenderView::viewTransitionRoot const): Deleted.
(WebCore::RenderView::setViewTransitionRoot): Deleted.
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::renderer const):

Canonical link: https://commits.webkit.org/295083@main
